File: daemon/src/services/promptBuilder.ts
/**
 * 3-layer prompt builder per HIGH-LEVEL-DESIGN.md §4.
 *
 * L1 — base: skill/skill.md (loaded once at daemon boot, cached)
 * L2 — context: project + worktree + mode context
 * L3 — rules: <project>/AGENTS.md or <project>/.vibe-station/rules.md (read at every spawn)
 *
 * Output: { systemPrompt, taskPrompt? }
 */
import { readFile } from "node:fs/promises";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import type { ProjectRecord, WorktreeRecord } from "../types.js";
import { worktreePath } from "./paths.js";

const here = dirname(fileURLToPath(import.meta.url));

// agent-system-prompt.md lives at ../assets/agent-system-prompt.md relative to this file,
// which resolves correctly under both:
//   compiled: dist/daemon/services/ → dist/daemon/assets/agent-system-prompt.md
//   vitest:   src/daemon/services/  → src/daemon/assets/agent-system-prompt.md
function assetPath(): string {
  return join(here, "..", "assets", "agent-system-prompt.md");
}

let cachedSkillMd: string | undefined;

async function loadSkillMd(): Promise<string> {
  if (cachedSkillMd !== undefined) return cachedSkillMd;
  try {
    const content = await readFile(assetPath(), "utf8");
    cachedSkillMd = content;
    return content;
  } catch {
    // Fallback if asset not found (should not happen in normal operation)
    cachedSkillMd = "# vibe-station Agent\n\nYou are a vibe-station-managed coding agent.";
    return cachedSkillMd;
  }
}

/** Force-reload the skill.md cache (useful for tests). */
export function _resetSkillCacheForTest(): void {
  cachedSkillMd = undefined;
}

export interface BuildPromptInput {
  project: ProjectRecord;
  worktree: WorktreeRecord;
  modeContext?: string;
  userPrompt?: string;
}

export interface BuildDirectPromptInput {
  project: ProjectRecord;
  modeContext?: string;
  userPrompt?: string;
}

export interface BuiltPrompt {
  systemPrompt: string;
  taskPrompt?: string;
}

/**
 * Build the layered prompt for an agent spawn.
 */
export async function buildPrompt(input: BuildPromptInput): Promise<BuiltPrompt> {
  const { project, worktree, modeContext, userPrompt } = input;

  // L1 — base skill
  const l1 = await loadSkillMd();

  // L2 — project + worktree + mode context
  const wtPath = worktreePath(project.id, worktree.id);
  const l2Lines: string[] = [
    "",
    "## Context",
    "",
    `**Your working directory (worktree):** ${wtPath}`,
    `> This is where you must read and write all files. Do NOT work in the project base directory.`,
    "",
    `**Project:** ${project.id}`,
    `**Project base directory:** ${project.absolutePath} — for reference only; do not edit files here`,
    `**Default branch:** ${project.defaultBranch}`,
    `**Worktree:** ${worktree.id}`,
    `**Branch:** ${worktree.branch}`,
    `**Base branch:** ${worktree.baseBranch} @ ${worktree.baseSha}`,
  ];

  if (worktree.sessions.length > 0) {
    l2Lines.push("", "**Sibling sessions in this worktree:**");
    for (const s of worktree.sessions) {
      l2Lines.push(`- ${s.id} (slot=${s.slot}, type=${s.type}, state=${s.lifecycle.state})`);
    }
  }

  if (modeContext) {
    l2Lines.push("", "## Mode Instructions", "", modeContext);
  }

  const l2 = l2Lines.join("\n");

  // L3 — project-level rules (AGENTS.md or .vibe-station/rules.md)
  const l3 = await readProjectRules(project.absolutePath);

  const systemPrompt = [l1, l2, ...(l3 ? [l3] : [])].join("\n");

  return {
    systemPrompt,
    taskPrompt: userPrompt || undefined,
  };
}

/**
 * Build prompt for a direct session (no worktree).
 * Simpler context — just project, no branch/worktree info.
 */
export async function buildDirectPrompt(input: BuildDirectPromptInput): Promise<BuiltPrompt> {
  const { project, modeContext, userPrompt } = input;

  // L1 — base skill
  const l1 = await loadSkillMd();

  // L2 — project context (no worktree)
  const l2Lines: string[] = [
    "",
    "## Context",
    "",
    `**Your working directory:** ${project.absolutePath}`,
    `> This is a direct session running in the project directory (no worktree isolation).`,
    "",
    `**Project:** ${project.id}`,
  ];

  // Only show defaultBranch for git projects
  if (project.isGit && project.defaultBranch) {
    l2Lines.push(`**Default branch:** ${project.defaultBranch}`);
    l2Lines.push("", "> Note: You are editing files directly in the project directory.");
    l2Lines.push("> Changes will affect the current working tree (no branch isolation).");
  } else {
    l2Lines.push("", "> This is a non-git project. No version control is active.");
  }

  if (modeContext) {
    l2Lines.push("", "## Mode Instructions", "", modeContext);
  }

  const l2 = l2Lines.join("\n");

  // L3 — project-level rules
  const l3 = await readProjectRules(project.absolutePath);

  const systemPrompt = [l1, l2, ...(l3 ? [l3] : [])].join("\n");

  return {
    systemPrompt,
    taskPrompt: userPrompt || undefined,
  };
}

async function readProjectRules(projectPath: string): Promise<string | null> {
  const candidates = [
    join(projectPath, "AGENTS.md"),
    join(projectPath, ".vibe-station", "rules.md"),
  ];
  for (const candidate of candidates) {
    try {
      return await readFile(candidate, "utf8");
    } catch {
      // file not found — try next
    }
  }
  return null;
}

File: rust/vst-agents/src/prompt_builder.rs
//! 3-layer prompt builder per HIGH-LEVEL-DESIGN.md §4 (ports
//! `services/promptBuilder.ts`).
//!
//! L1 — base: `daemon/src/assets/agent-system-prompt.md` (embedded at compile
//! time, cached) + the Rich-Chat-only subagent fragment when `rich_chat`.
//! L2 — context: project + worktree + mode context.
//! L3 — rules: `<project>/AGENTS.md` or `<project>/.vibe-station/rules.md`.
//!
//! The L1/subagent assets are embedded via `include_str!` so the crate is
//! self-contained and tests are deterministic, while keeping the daemon's
//! asset files the single source of truth.

use std::sync::Mutex;

use vst_types::{LifecycleState, ProjectRecord, SessionType, WorktreeRecord};

// The daemon asset files, embedded once at compile time.
const SKILL_MD: &str = include_str!("../../../daemon/src/assets/agent-system-prompt.md");
const SUBAGENT_MD: &str = include_str!("../../../daemon/src/assets/agent-subagent-richchat.md");

static CACHED_SKILL_MD: Mutex<Option<String>> = Mutex::new(None);
static CACHED_SUBAGENT_MD: Mutex<Option<String>> = Mutex::new(None);

fn load_skill_md() -> String {
    let mut cache = CACHED_SKILL_MD.lock().unwrap();
    if let Some(c) = cache.as_ref() {
        return c.clone();
    }
    let content = SKILL_MD.to_string();
    *cache = Some(content.clone());
    content
}

fn load_subagent_md() -> String {
    let mut cache = CACHED_SUBAGENT_MD.lock().unwrap();
    if let Some(c) = cache.as_ref() {
        return c.clone();
    }
    let content = SUBAGENT_MD.to_string();
    *cache = Some(content.clone());
    content
}

/// Force-reload the skill cache (useful for tests).
pub fn reset_skill_cache_for_test() {
    *CACHED_SKILL_MD.lock().unwrap() = None;
    *CACHED_SUBAGENT_MD.lock().unwrap() = None;
}

/// Input to `build_prompt` (worktree context).
#[derive(Clone)]
pub struct BuildPromptInput {
    pub project: ProjectRecord,
    pub worktree: WorktreeRecord,
    pub mode_context: Option<String>,
    pub user_prompt: Option<String>,
    pub rich_chat: bool,
}

/// Input to `build_direct_prompt` (direct context, no worktree).
#[derive(Clone)]
pub struct BuildDirectPromptInput {
    pub project: ProjectRecord,
    pub mode_context: Option<String>,
    pub user_prompt: Option<String>,
    pub rich_chat: bool,
}

/// Output of either builder.
#[derive(Clone, Debug)]
pub struct BuiltPrompt {
    pub system_prompt: String,
    pub task_prompt: Option<String>,
}

/// Build the layered prompt for a worktree agent spawn.
pub fn build_prompt(input: &BuildPromptInput) -> BuiltPrompt {
    let l1 = if input.rich_chat {
        format!("{}\n{}", load_skill_md(), load_subagent_md())
    } else {
        load_skill_md()
    };

    let wt_path = format!(
        "{}/projects/{}/worktrees/{}",
        crate::home::home_dir().display(),
        input.project.id,
        input.worktree.id
    );
    let mut l2_lines: Vec<String> = vec![
        String::new(),
        "## Context".into(),
        String::new(),
        format!("**Your working directory (worktree):** {wt_path}"),
        "> This is where you must read and write all files. Do NOT work in the project base directory.".into(),
        String::new(),
        format!("**Project:** {}", input.project.id),
        format!("**Project base directory:** {} — for reference only; do not edit files here", input.project.absolute_path),
        format!("**Default branch:** {}", input.project.default_branch.as_deref().unwrap_or("main")),
        format!("**Worktree:** {}", input.worktree.id),
        format!("**Branch:** {}", input.worktree.branch),
        format!("**Base branch:** {} @ {}", input.worktree.base_branch, input.worktree.base_sha),
    ];

    if !input.worktree.sessions.is_empty() {
        l2_lines.push(String::new());
        l2_lines.push("**Sibling sessions in this worktree:**".into());
        for s in &input.worktree.sessions {
            let ty = session_type_str(s.r#type);
            let state = lifecycle_state_str(s.lifecycle.state);
            l2_lines.push(format!(
                "- {} ({}, type={}, state={})",
                s.id,
                if s.is_main { "main" } else { ty },
                ty,
                state
            ));
        }
    }

    if let Some(mode_context) = &input.mode_context {
        l2_lines.push(String::new());
        l2_lines.push("## Mode Instructions".into());
        l2_lines.push(String::new());
        l2_lines.push(mode_context.clone());
    }

    let l2 = l2_lines.join("\n");
    let l3 = read_project_rules(&input.project.absolute_path);

    let mut system_prompt = l1;
    system_prompt.push('\n');
    system_prompt.push_str(&l2);
    if let Some(rules) = l3 {
        system_prompt.push('\n');
        system_prompt.push_str(&rules);
    }

    BuiltPrompt {
        system_prompt,
        task_prompt: input.user_prompt.clone().filter(|s| !s.is_empty()),
    }
}

/// Build the layered prompt for a direct (no-worktree) agent spawn.
pub fn build_direct_prompt(input: &BuildDirectPromptInput) -> BuiltPrompt {
    let l1 = if input.rich_chat {
        format!("{}\n{}", load_skill_md(), load_subagent_md())
    } else {
        load_skill_md()
    };

    let mut l2_lines: Vec<String> = vec![
        String::new(),
        "## Context".into(),
        String::new(),
        format!(
            "**Your working directory:** {}",
            input.project.absolute_path
        ),
        "> This is a direct session running in the project directory (no worktree isolation)."
            .into(),
        String::new(),
        format!("**Project:** {}", input.project.id),
    ];

    if input.project.is_git && input.project.default_branch.is_some() {
        l2_lines.push(format!(
            "**Default branch:** {}",
            input.project.default_branch.as_deref().unwrap()
        ));
        l2_lines.push(String::new());
        l2_lines.push("> Note: You are editing files directly in the project directory.".into());
        l2_lines
            .push("> Changes will affect the current working tree (no branch isolation).".into());
    } else {
        l2_lines.push(String::new());
        l2_lines.push("> This is a non-git project. No version control is active.".into());
    }

    if let Some(mode_context) = &input.mode_context {
        l2_lines.push(String::new());
        l2_lines.push("## Mode Instructions".into());
        l2_lines.push(String::new());
        l2_lines.push(mode_context.clone());
    }

    let l2 = l2_lines.join("\n");
    let l3 = read_project_rules(&input.project.absolute_path);

    let mut system_prompt = l1;
    system_prompt.push('\n');
    system_prompt.push_str(&l2);
    if let Some(rules) = l3 {
        system_prompt.push('\n');
        system_prompt.push_str(&rules);
    }

    BuiltPrompt {
        system_prompt,
        task_prompt: input.user_prompt.clone().filter(|s| !s.is_empty()),
    }
}

fn read_project_rules(project_path: &str) -> Option<String> {
    let candidates = [
        std::path::Path::new(project_path).join("AGENTS.md"),
        std::path::Path::new(project_path)
            .join(".vibe-station")
            .join("rules.md"),
    ];
    for candidate in &candidates {
        if let Ok(content) = std::fs::read_to_string(candidate) {
            return Some(content);
        }
    }
    None
}

fn session_type_str(t: SessionType) -> &'static str {
    match t {
        SessionType::Agent => "agent",
        SessionType::Terminal => "terminal",
    }
}

fn lifecycle_state_str(s: LifecycleState) -> &'static str {
    match s {
        LifecycleState::NotStarted => "not_started",
        LifecycleState::Working => "working",
        LifecycleState::Idle => "idle",
        LifecycleState::WaitingForHuman => "waiting_for_human",
        LifecycleState::Done => "done",
        LifecycleState::Exited => "exited",
        LifecycleState::Drafting => "drafting",
    }
}
